File: DS/Trees/Binary tree/search tree/PYTHON/BST_implementation.py
#class

class BstNode:
    def __init__(self, num):
        self.data = num
        self.left = None
        self.right = None

#insert function

def Insert(pointer, data):
    if pointer is None:
        pointer = BstNode(data)

    elif data <= pointer.data:
        pointer.left = Insert(pointer.left, data)

    else:
        pointer.right = Insert(pointer.right, data)

    return pointer

#search function

def Search(pointer, data):
    if pointer is None:
        return False
    if data == pointer.data:
        return True

    elif data < pointer.data:
        return Search(pointer.left, data)

    else:
        return Search(pointer.right, data)

if __name__ == "__main__":

    root = None

    root = Insert(root, 10)
    root = Insert(root, 8)
    root = Insert(root, 9)
    root = Insert(root, 7)
    root = Insert(root, 25)
    root = Insert(root, 30)
    root = Insert(root, 15)

    num = int(input("Enter a number: "))

    print("Found" if Search(root, num) else "Not found")
   
# input: 8
# output: Found

File: DS/Trees/Binary tree/search tree/CPP/BST_implementation.cpp
#include <bits/stdc++.h>
using namespace std;

// class for node

class bstNode
{
public:
    int data;
    bstNode* left;
    bstNode* right;

    bstNode(int num){
        data = num;
        left = NULL;
        right = NULL;
    }

};

//insert function

bstNode* insert(bstNode* pointer, int data){
    if(pointer == NULL){
        pointer = new bstNode(data);
    }

    else if(data <= pointer->data){
        pointer->left = insert(pointer->left, data);
    }

    else{
        pointer->right = insert(pointer->right, data);
    }

    return pointer;
}

//search function

bool search(bstNode* pointer, int data){
    if(pointer==NULL) return false;
    if(data == pointer->data) 
    return true;

    else if(data < pointer->data)
    return search(pointer->left, data);

    else
    return search(pointer->right, data);
} 

//main

int main(){
    bstNode* root = NULL;
    root = insert(root, 10);
    root = insert(root, 25);
    root = insert(root, 8);
    root = insert(root, 9);
    root = insert(root, 7);
    root = insert(root, 15);
    root = insert(root, 30);

    int num;
    cout<<"Enter a number: ";
    cin>>num;


    if(search(root, num) == true){
        cout<<"\nFound";
    }

    else{
        cout<<"\nNot found";
    }
}

// input: 8
// output: Found
